app/routes.py

```diff
@@ -1,2 +1,273 @@
+    # Update a task
+    elif request.method == "PUT":
+        if task is None:
+            return make_response(f"Task #{task_id} not found"), 404
+        form_data = request.get_json()
+
+        if "title" not in form_data or "description" not in form_data or "completed_at" not in form_data:
+            return {"details": "Invalid data"}, 400
+
+        if not "completed_at" in form_data or not form_data["completed_at"]:
+            is_complete = False
+        else:
+            is_complete = True            
+        task.title = form_data["title"]
+        task.description = form_data["description"]
+        task.completed_at = form_data["completed_at"]
+
+        db.session.commit()
+        return {
+            "task": {
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description, 
+                "is_complete": is_complete
+            }
+        }, 200
+
+    # Delete a task
+    elif request.method == "DELETE":
+        if not task:
+            return "", 404
+        db.session.delete(task)
+        db.session.commit()
+        return {
+        "details": f"Task {task_id} \"{task.title}\" successfully deleted"
+        }, 200
+            
+
+# modify complete task
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"], strict_slashes = False)
+def mark_task_complete(task_id):
+    task = Task.query.get(task_id)
+
+    if not task: # this is the task check (wash my car)
+        return make_response(f"Task #{task_id} not found"), 404
+    task.completed_at = datetime.utcnow()
+    db.session.commit()
+
+    base_path = "https://slack.com/api/chat.postMessage"
+    slack_tocken = os.environ.get("SLACK_API_KEY")
+    query_params = {
+        "channel": "task-notifications",
+        "text": f"Someone just completed the {task.title}"
+    }
+    headers = {
+        "Authorization": f"Bearer {slack_tocken}"
+    }
+    requests.post(base_path, params=query_params, headers=headers)
+
+    response_body = {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": True # this is always true bc we are always setting completed at
+        }
+    }
+    return jsonify(response_body), 200
+
+# modify incomplete task
+@task_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def make_task_incomplete(task_id):
+    task = Task.query.get(task_id)
+
+    if task is None:
+        return make_response(f"Task #{task_id} not found"), 404
+    else:
+        task.completed_at = None
+        db.session.commit()
+    response_body = {
+        "task": {
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete":  False
+        }
+    }
+    return jsonify(response_body), 200
+
+
+# goals route
+@goals_bp.route("", methods = ["POST", "GET"], strict_slashes = False)
+def handle_goals():
```

Comment:
Similar feedback about splitting these functions, and moving validation and dictionary-handling logic around that I made for `Task` will also apply to `Goal`. These are common themes for any model with CRUD operations.